### PR TITLE
feat(NL): Use NED.nl as source for solar data

### PR DESCRIPTION
**Pull request by VIKTORVAV99**
## Issue

ENTSO-E don't have any good solar data for NL.

## Description

This PR adds a parser that gets the solar values from NED.nl and the production data from ENTSO-E, it then updates the ENTSO-E data with the data from NED.nl, it does this for both production and forecasts.

>[!IMPORTANT]
>We need to change the estimation method for NL in conjunction with merging this and then refetch and reprocess all the historical data.

>[!WARNING]
>We need to add a API KEY for NED before this is merged!

### Double check

- [x] I have tested my parser changes locally with `poetry run test_parser "zone_key"`
- [x] I have run `pnpx prettier@2 --write .` and `poetry run format` in the top level directory to format my changes.

### CODE REVIEW

File: electricitymap/contrib/lib/models/event_lists.py [hunk 1]
```diff
@@ -302,20 +302,36 @@ def update_production_breakdowns(
...
-            else:
-                production_breakdowns.append(
+                updated_production_breakdowns.append(
+                    updated_event.zoneKey,
+                    updated_event.datetime,
+                    updated_event.source,
+                    updated_event.production,
+                    updated_event.storage,
+                    updated_event.sourceType,
+                )
```
Comment (VIKTORVAV99) [inline electricitymap/contrib/lib/models/event_lists.py:332]: This will be simplified in another PR and abstracted to a helper function.

File: electricitymap/contrib/lib/models/events.py [hunk 2]
```diff
@@ -655,20 +655,19 @@ def _update(
...
-                f"Cannot update events from different sources: {event.source} and {new_event.source}"
-            )
         if event.sourceType != new_event.sourceType:
             raise ValueError(
                 f"Cannot update events from different source types: {event.sourceType} and {new_event.sourceType}"
             )
         production_mix = ProductionMix._update(event.production, new_event.production)
         storage_mix = StorageMix._update(event.storage, new_event.storage)
+        source = ", ".join(
+            set(event.source.split(", ")) | set(new_event.source.split(", "))
```
Comment (VIKTORVAV99) [inline electricitymap/contrib/lib/models/events.py:665]: This ensures that if we where to update and already updated list that now contain two sources these are deduplicated.

File: parsers/NED.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        mix = ProductionMix()
+        for data in data_dict:
+            clean_type = int(data["type"].split("/")[-1])
+            if clean_type in TYPE_MAPPING:
+                mix.add_value(
+                    TYPE_MAPPING[clean_type],
+                    _kwh_to_mw(data["volume"]),
+                )
+            else:
+                logger.warning(f"Unknown type: {clean_type}")
```
Comment (VIKTORVAV99) [inline parsers/NED.py:133]: This is a bit overkill for just solar but makes it easy to add other modes as well in the future, like wind for example.

File: electricitymap/contrib/lib/models/event_lists.py [hunk 4]
```diff
@@ -302,20 +302,36 @@ def update_production_breakdowns(
         production_breakdowns: "ProductionBreakdownList",
         new_production_breakdowns: "ProductionBreakdownList",
         logger: Logger,
+        matching_timestamps_only: bool = False,
```
Comment (pierresegonne) [inline electricitymap/contrib/lib/models/event_lists.py:305]: Maybe worth adding a comment at this level to explain what this parameter does?

File: electricitymap/contrib/lib/models/event_lists.py [hunk 5]
```diff
@@ -324,7 +340,19 @@ def update_production_breakdowns(
                     new_event.sourceType,
                 )
 
-        return production_breakdowns
+        if matching_timestamps_only is False:
+            for existing_event in production_breakdowns.events:
```
Comment (pierresegonne) [inline electricitymap/contrib/lib/models/event_lists.py:344]: Same here, could be nice to have an helper function for that

File: parsers/NED.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        else NedClassification.MEASURED.value,
+        "activity": NedActivity.PRODUCTION.value,
+        "validfrom[before]": (target_datetime + timedelta(days=2 if forecast else 1))
+        .date()
+        .isoformat(),
+        "validfrom[after]": (target_datetime - timedelta(days=0 if forecast else 1))
+        .date()
+        .isoformat(),
+    }
+    headers = {"X-AUTH-TOKEN": get_token("NED_KEY"), "accept": "application/json"}
```
Comment (pierresegonne) [inline parsers/NED.py:90]: We're using `${SOURCE}_TOKEN` for all (except the EIA...)

File: parsers/NED.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    response = requests.get(URL, params=params, headers=headers)
+    if not response.ok:
+        raise ParserException(
+            parser="NED.py",
+            message=f"Failed to fetch NED data: {response.status_code}, err: {response.text}",
+        )
+    return response.json()
+
+
+def format_data(json: Any, logger: Logger, forecast: bool = False):
```
Comment (pierresegonne) [inline parsers/NED.py:100]: Can you add a type for the return value pls?

File: parsers/NED.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+    ENTSOE_raw_data = ENTSOE_query_production(
+        ENTSOE_DOMAIN_MAPPINGS[zone_key], session, target_datetime=target_datetime
+    )
+    if ENTSOE_raw_data is None:
+        raise ParserException(
+            parser="NED.py",
+            message="Failed to fetch ENTSOE data",
+            zone_key=zone_key,
+        )
+    ENTSOE_parsed_data = ENTSOE_parse_production(
```
Comment (pierresegonne) [inline parsers/NED.py:165]: Could be extracted to its own function?

**Review by pierresegonne — APPROVED** [review]
Mostly minor comments to make code more readable - nothing blocking, thanks for doing it!

File: electricitymap/contrib/lib/models/event_lists.py (hunk 5 above)
Comment (VIKTORVAV99) [reply electricitymap/contrib/lib/models/event_lists.py:344]: Will add that in another PR so this don't require another review.

**Review by pierresegonne — APPROVED** [review]

**Comment by mhooi** [comment]
First of all thank you for implementing this major improvement so fast. 

@VIKTORVAV99: you also wrote it in one of your comments but next to solar I think it is good to add also wind energy. Currently only Wind generation from Sea is reported correctly on ENTSO-E. Wind generation from land is not included but is luckily in cluded in the NED.NL API. 

Looking for example at Electricity maps to the wind data on the 28th of April 11h00 there is only a value of 2.86 GW. 
When we look at [https://co2monitor.nl/en/energysources/history](url) we see the same 2.8GW for wind at see but there is also roughly 5 GW reported under wind from land. SO Electricity maps is highly underestimating the wind generation. 

I hope after having implemented the code the addition of Wind energy is relatively easy?

**Comment by mhooi** [comment]
Ofcourse, no hurry. I'm just a bit eager to have finally "correct" or alligned data for the Netherlands :p 2.8 vs 7.8 GW is for example a big difference.

**Comment by VIKTORVAV99** [comment]
Agreed!

I will open a internal issue about it so we can keep track of it 🙂
